pkg/appgateserver/sdkadapter/sdk.go

```diff
@@ -126,12 +199,28 @@ func (shannonSDK *ShannonSDK) SendRelay(
 }
 
 // GetSessionSupplierEndpoints returns the current session's supplier endpoints
-// for the given appAddress and serviceId.
+// for the given appAddress and serviceId, using a cached result if available.
 func (shannonSDK *ShannonSDK) GetSessionSupplierEndpoints(
 	ctx context.Context,
 	appAddress, serviceId string,
 ) (*shannonsdk.SessionFilter, error) {
-	currentHeight := shannonSDK.blockClient.LastBlock(ctx).Height()
+	// Use the cached block height
+	currentHeight, err := shannonSDK.GetCachedHeight(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	key := cacheKey(appAddress, serviceId, currentHeight)
+
+	// Check if result is cached
+	sessionCacheInstance.mu.RLock()
+	if result, found := sessionCacheInstance.cache[key]; found {
+		sessionCacheInstance.mu.RUnlock()
+		return result.filter, nil
+	}
+	sessionCacheInstance.mu.RUnlock()
+
+	// Fetch session from the session client if not cached
 	session, err := shannonSDK.sessionClient.GetSession(ctx, appAddress, serviceId, currentHeight)
```

Comment:
This (and the height request) rendered appgate server useless under load, as even large fullnode couldn't handle all the requests. There's 100% better way to cache this.